File: reference/go/aep/envelope.go
package aep

import (
	"time"
)

var deliveryModes = map[string]bool{
	"best_effort":   true,
	"at_least_once": true,
	"replayable":    true,
}

func ValidateEnvelope(value map[string]any) []string {
	var errors []string

	if value == nil {
		return []string{"event must be a JSON object"}
	}

	requireString(value, "aep_version", &errors)
	requireString(value, "id", &errors)
	requireString(value, "type", &errors)
	requireString(value, "source", &errors)
	requireString(value, "created_at", &errors)

	if _, ok := value["payload"]; !ok {
		errors = append(errors, "payload is required")
	}

	if typ, ok := value["type"].(string); ok && !IsStandardEventType(typ) {
		errors = append(errors, "type is not in the standard draft registry: "+typ)
	}

	if v, ok := value["aep_version"].(string); ok && v != "0.1" {
		errors = append(errors, "unsupported protocol version: "+v)
	}

	if ts, ok := value["created_at"].(string); ok {
		if _, err := time.Parse(time.RFC3339, ts); err != nil {
			errors = append(errors, "created_at must be an ISO-compatible timestamp")
		}
	}

	if delivery, ok := value["delivery"]; ok {
		validateDelivery(delivery, &errors)
	}

	if typ, _ := value["type"].(string); typ == "subscription.requested" {
		if payload, ok := value["payload"].(map[string]any); ok {
			validateSubscriptionPayload(payload, &errors)
		}
	}

	return errors
}

func requireString(value map[string]any, field string, errors *[]string) {
	s, ok := value[field].(string)
	if !ok || s == "" {
		*errors = append(*errors, field+" must be a non-empty string")
	}
}

func validateDelivery(delivery any, errors *[]string) {
	d, ok := delivery.(map[string]any)
	if !ok {
		*errors = append(*errors, "delivery must be an object when present")
		return
	}
	if mode, ok := d["mode"].(string); ok && !deliveryModes[mode] {
		*errors = append(*errors, "delivery.mode must be one of: best_effort, at_least_once, replayable")
	}
}

func validateSubscriptionPayload(payload map[string]any, errors *[]string) {
	if types, ok := payload["types"]; ok && !isStringOrSlice(types) {
		*errors = append(*errors, "subscription payload types must be a string or string array")
	}
	fields := []string{"source", "target", "topic", "session_id", "conversation_id", "task_id"}
	for _, f := range fields {
		if v, ok := payload[f]; ok && !isStringOrSlice(v) {
			*errors = append(*errors, "subscription payload "+f+" must be a string or string array")
		}
	}
}

func isStringOrSlice(v any) bool {
	if _, ok := v.(string); ok {
		return true
	}
	arr, ok := v.([]any)
	if !ok {
		return false
	}
	for _, item := range arr {
		if _, ok := item.(string); !ok {
			return false
		}
	}
	return true
}

File: implementations/java/src/main/java/com/axisrobo/harmovela/event/envelope/Envelope.java
package com.axisrobo.harmovela.event.envelope;

import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class Envelope {
    private Envelope() {}

    private static final Set<String> DELIVERY_MODES = Set.of("best_effort", "at_least_once", "replayable");
    private static final DateTimeFormatter ISO = DateTimeFormatter.ISO_DATE_TIME;

    public static final Set<String> STANDARD_TYPES = Set.of(
        "session.opened", "session.ready", "session.heartbeat", "session.closed", "session.error",
        "capabilities.requested", "capabilities.declared", "capabilities.changed",
        "subscription.requested", "subscription.created", "subscription.rejected",
        "subscription.cancelled", "subscription.expired",
        "event.acknowledged", "event.rejected", "event.redelivered", "event.dead_lettered",
        "task.submitted", "task.accepted", "task.started", "task.progress", "task.blocked",
        "task.output", "task.completed", "task.failed", "task.cancelled", "task.timed_out",
        "task.cancel.requested",
        "memory.updated", "memory.fact.added", "memory.fact.updated", "memory.fact.invalidated",
        "memory.episode.stored", "memory.preference.updated", "memory.constraint.updated",
        "memory.summary.ready", "memory.retrieval.ready",
        "context.updated", "context.invalidated",
        "context.snapshot.requested", "context.snapshot.ready",
        "context.retrieval.started", "context.retrieval.completed", "context.retrieval.failed",
        "belief.revised", "belief.conflict.detected",
        "provenance.attestation.added", "provenance.attestation.revoked", "provenance.chain.truncated",
        "interruption.requested", "interruption.acknowledged", "interruption.saved",
        "interruption.resumed", "interruption.cancelled",
        "compensation.requested", "compensation.completed",
        "freshness.expired",
        "delegation.requested", "delegation.accepted", "delegation.rejected",
        "delegation.escalated", "delegation.handoff.completed",
        "adaptation.outcome.correlated",
        "adaptation.goal.created", "adaptation.goal.updated",
        "adaptation.goal.achieved", "adaptation.goal.abandoned",
        "adaptation.cost.exceeded",
        "adaptation.budget.established", "adaptation.budget.adjusted",
        "adaptation.budget.limit_exceeded", "adaptation.budget.exhausted",
        "command.requested", "command.accepted", "command.rejected",
        "command.completed", "command.failed",
        "query.requested", "query.response", "query.rejected",
        "query.error",
        "capability.registered", "capability.updated", "capability.deprecated",
        "capability.composed", "capability.validated"
    );

    public static boolean isStandardEventType(String type) {
        return type != null && STANDARD_TYPES.contains(type);
    }

    public static List<String> validate(Map<String, Object> value) {
        var errors = new ArrayList<String>();
        if (value == null) return List.of("event must be a JSON object");

        requireString(value, "spec_version", errors);
        requireString(value, "id", errors);
        requireString(value, "type", errors);
        requireString(value, "source", errors);
        requireString(value, "created_at", errors);
        if (!value.containsKey("payload")) errors.add("payload is required");

        var type = (String) value.get("type");
        if (type != null && !isStandardEventType(type)) {
            errors.add("type is not in the standard draft registry: " + type);
        }
        var version = (String) value.get("spec_version");
        if (version != null && !"0.2".equals(version)) errors.add("unsupported protocol version: " + version);

        var ts = (String) value.get("created_at");
        if (ts != null) {
            try {
                ISO.parse(ts);
            } catch (DateTimeParseException e) {
                errors.add("created_at must be an ISO-compatible timestamp");
            }
        }
        if (value.containsKey("delivery")) validateDelivery(value.get("delivery"), errors);
        if ("subscription.requested".equals(type) && value.get("payload") instanceof Map<?, ?> payload) {
            validateSubscriptionPayload(payload, errors);
        }
        return errors;
    }

    private static void validateSubscriptionPayload(Map<?, ?> payload, List<String> errors) {
        if (payload.containsKey("types") && !isStringOrList(payload.get("types"))) {
            errors.add("subscription payload types must be a string or string array");
        }
        for (var field : List.of("source", "target", "topic", "session_id", "conversation_id", "task_id")) {
            if (payload.containsKey(field) && !isStringOrList(payload.get(field))) {
                errors.add("subscription payload " + field + " must be a string or string array");
            }
        }
    }

    private static void validateDelivery(Object delivery, List<String> errors) {
        if (!(delivery instanceof Map<?, ?> values)) {
            errors.add("delivery must be an object when present");
            return;
        }
        var mode = values.get("mode");
        if (mode instanceof String value && !DELIVERY_MODES.contains(value)) {
            errors.add("delivery.mode must be one of: best_effort, at_least_once, replayable");
        }
    }

    private static void requireString(Map<String, Object> value, String field, List<String> errors) {
        var fieldValue = value.get(field);
        if (!(fieldValue instanceof String string) || string.isEmpty()) {
            errors.add(field + " must be a non-empty string");
        }
    }

    private static boolean isStringOrList(Object value) {
        if (value instanceof String) return true;
        return value instanceof List<?> list && list.stream().allMatch(item -> item instanceof String);
    }
}
